### DISCUSSION TITLE
stm32f411 black pill not idenitified

### DISCUSSION THREAD

**Discussion by mogplus88** [micropython/micropython#12692] [STM32 / Pyboard] [answered]
Hi All,
Apologies if this is the second time I've posted this. I thought I'd already posted but I can't find it, so maybe I forgot to hit the "submit" button. Wouldn't be the first time...
My black pill is shown when I do a `lsusb` (Linux Mint) like this `Bus 001 Device 007: ID 0483:572a STMicroelectronics WeAct Studio HID Bootloader`, but when I do `ls /dev/tty*` it does not appear. So needless to say Thonny can't find it. 
What am I missing?

**Comment by andrewleech** [comment]
The device sounds like it's in dfu bootloader mode, probably hasn't been flashed with firmware yet? Have you found / built a copy of micropython that's compatible with that board?

**Reply by mogplus88** [reply]
I thought it must have something installed on it as one of the leds fades up and down when I plug it in. 
Not sure about a micropython build for it. It has an stm32f411 processor so I'm guessing the Pyboard build would work. Or there's an stm discovery f411 board, but it looks a bit vintage...

**Reply by andrewleech** [reply]
It sounds like it might have a "custom" bootloader already installed if it has weact in the name on device manager.

If you search PR's here you should find one that matches, or yeah pyboard-lite might be close enough.

**Comment by nspsck** [comment]
Hi,

This should probably solve your issue:
[https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython](https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython)

You can simply flash the pre-built firmware using STM32CubeProgrammer, or you can just build your own firmware following the instructions. It supports micropython 1.20.0 and upwards.

**Reply by mogplus88** [reply]
@nspsck thanks for the link. I tried the instructions there and after a few false starts got it to work. But my main problem is actually deploying the firmware to the board. The```make ... deploy``` command doesn't work as I get an error message saying module usb is missing.
```
ian@Mintian:~/micropython/ports/stm32$ make -j LTO=1 BOARD=WEACTF411CE BOARD_VARIANT=DP deploy
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Writing build-WEACTF411CE-DP/firmware.dfu to the board
Traceback (most recent call last):
  File "/home/ian/micropython/ports/stm32/../../tools/pydfu.py", line 22, in <module>
    import usb.core
ModuleNotFoundError: No module named 'usb'
make: *** [Makefile:604: deploy] Error 1
```
I tried STM32CubeProgrammer but I couldn't get it to connect to the board. I don't really know how to work it though, so that could have been a problem between keyboard and seat. But it may be because there is no /dev/tty* for it. ```dfu-util``` can't find it either. 
```lsusb``` will display an entry for it though, but not if it's in dfu mode, only if I just plug it in normally.
It can't be a dud cable because ```lsusb``` wouldn't work.
Maybe it's a dud board.

:-(

**Reply by rkompass** [reply]
> ModuleNotFoundError: No module named 'usb'

indicates that [pydfu.py](https://github.com/micropython/micropython/blob/master/tools/pydfu.py)  was not able to import the usb module. You might look for this module, install it (with `pip install pyusb`) and try again.

In the description of pydfu it says:
> This module implements enough functionality to program the STM32F4xx over DFU, without requiring dfu-util.

So as an alternative you may install dfu-util and use that. That was what I did :-) There are many descriptions how to deploy firmware to the pyboard with dfu-util that apply here.

**Reply by nspsck** [reply]
To enter the DFU you have to hold down the boot button during boot/reboot. The weact hid bootloader is , as far as I know, not actually a standard DFU, it's designed to be used with their software together. Make sure you are using the ST bootloader, by checking with `lsusb`， it should give you something like this:

```shell
paul@TNsprier:~/st7789_dev/mpy/micropython/ports/stm32$ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 002: ID 0483:df11 STMicroelectronics STM Device in DFU Mode
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

If you were using windows subsystem for linux, you probably need to run the flash command in `sudo`, or just use `chown` to make sure that device is exposed to you.

Without `sudo`:
```shell
paul@TNsprier:~/st7789_dev/mpy/micropython/ports/stm32/build-WEACTF411CE$ dfu-util --alt 0 -D firmware.dfu
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Match vendor ID from file: 0483
Match product ID from file: df11
dfu-util: Cannot open DFU device 0483:df11
dfu-util: No DFU capable USB device available
```

With `sudo` or you are sure that you have the access to that device:
```shell
paul@TNsprier:~/st7789_dev/mpy/micropython/ports/stm32/build-WEACTF411CE$ sudo dfu-util --alt 0 -D firmware.dfu
[sudo] password for paul:
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Match vendor ID from file: 0483
Match product ID from file: df11
Opening DFU capable USB device...
ID 0483:df11
Run-time device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Setting #0 ...
Determining device status: state = dfuERROR, status = 10
dfuERROR, clearing status
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
file contains 1 DFU images
parsing DFU image 1
image for alternate setting 0, (2 elements, total size = 303784)
parsing element 1, address = 0x08000000, size = 408
Download        [=========================] 100%          408 bytes
Download done.
parsing element 2, address = 0x08030000, size = 303360
Download        [=========================] 100%       303360 bytes
Download done.
done parsing DfuSe file
```

And just in case, can you link where you got these boards from? Some aliexpress store sells pirated clones and to my experiences with those: they are not reliable at all. Some can not enter dfu properly, some's power supply (LDO) just stopped working.

**Reply by davefes** [reply]
Got the WEACTF411CE version running.   I have some W25Q32 (4MB) so looked at your WEACTF411CE8MB folder.

Would I just change line 38 in mpconfigboard.h to (32 * 1024 * 1024) and nothing else.  Minor point ... I think the comment in your code should say // 64M Mbit (8 MByte)

Thank you for posting the guide.

**Reply by nspsck** [reply]
Great! Glad that did help!

> I think the comment in your code should say // 64M Mbit (8 MByte)

Thank you for pointing that out! It's changed now!

**Reply by davefes** [reply]
> Would I just change line 38 in mpconfigboard.h to (32 * 1024 * 1024) and nothing else?

Thank you.

**Reply by nspsck** [reply]
Oh sorry, I wasn't aware of that question.

Yes, in case you have soldered that on a blackpill, then that is the only thing you need to do.

**Reply by davefes** [reply]
My fault ... I didn't put a ? mark on the original question.  I will put the chip on the board when I get the proper chip capacitors for C15.

**Reply by mogplus88** [reply]
Thanks for all the responses. 
@npsck ```lsusb``` gives me ```STMicroelectronics WeAct Studio HID Bootloader```. If I then put it into dfu mode by holding the boot button while pressing and releasing the reset button, lsusb shows nothing at all. The only thing that happens is the blue led turns off.  So maybe you're right about a non-standard bootloader.
I installed pyusb, but make...deploy still didn't work.
I tried dfu-util anyway, with and without sudo (I use Linux Mint) but in both cases it told me there was no dfu device available. 
I found a website that describes how to program the board with Arduino, and I know Arduino can burn a bootloader. but I haven't had a chance yet to try it. I'll report on that tomorrow.
:-\

**Reply by davefes** [reply]
I found that I had to try `dfu-util -l` several times.  Also, it seems that I need to plug-in, unplug and plug-in several times to get the board mounted.  Just realised that if I wait 12-15 seconds the board will mount.  Maybe this relates to my problem:
https://github.com/orgs/micropython/discussions/12758

Is yours an official WeAct board?  Mine appear to be pirated.

lsusb gives me `Bus 001 Device 017: ID f055:9800 WeActStudio BlackPill v3.1 Comm Port (FS Mode)` using @nspsck general image, which I built.

BTW, I got deploy to work following:
https://github.com/mcauser/WEACT_F411CEU6

**Reply by davefes** [reply]
@mogplus88, I have also noticed that I can modify the files on the desktop, because the device is mounted as well as in rshell.  This is different to the way I do things on the ESP32 and the PICO.  I wonder about your suggestion that maybe we need to be using the same bootloader and then we might stand a chance of moving forward.

**Reply by davefes** [reply, marked as answer]
It seems to be critical getting the timing right:

> This board can be flashed using DFU. To put the board in DFU mode, disconnect USB, connect A9 to A10 and reconnect USB. Press and hold BOOT0. Press and release RESET. Wait 0.5 seconds. Release BOOT0.

I think the critical part is how long you hold NRST in ... try 1 second

**Reply by mogplus88** [reply]
Good grief! It's only bloody worked! Held down the reset button for a count of five, and lsusb popped up the board in dfu mode. Tried the make...deploy, and it worked! I can now program the board with Thonny.
Many thanks to all who took the time to offer advice and assistance. And @davefes for the solution that did it.
:-)

**Reply by davefes** [reply]
For a 2nd board that I couldn't deploy to I followed this:
https://github.com/dhylands/upy-examples/blob/master/mass-erase.sh
and now have two boards working.

**Reply by mogplus88** [reply]
I ordered another one a couple of days ago as I thought the one I had was probably a dud, especially if it was a clone. I think it might be, as when I finally got it working I tried the blinky on it, and found that the machine.Pin.on() and machine.Pin.off() commands are backwards. off() turns the led on and on() turns it off. Bit weird. Anyway I ordered the new one from a different supplier, so maybe it will be better. 
Good to know about the mass erase though.
;-)

**Reply by davefes** [reply]
Is your LED connected to the pin and ground?

**Reply by mogplus88** [reply]
It's the builtin led on the board. Pin C13.

**Reply by davefes** [reply]
Looking at the schematic that LED is connected between 3V3 and pin13, so when pin13 is high the LED will be off.

**Reply by nspsck** [reply]
You should use pyb.LED to control the led. 

Please try:

```python
led=pyb.LED(1) # there's only one
led.on()
led.off()
```

And @davefes is right about the schematic. And `pyb.LED` is adjusted hence. But because you actually can get access to `PC13`, (not as a pwm-pin but a digital pin) it's only wise to keep `machine.Pin` output the correct signal.

Also here's a few helpful "commands" you should get fimilar with:

```python
help('modules') # Print out all the available module names.
dir(module_name) # Print out all the functions/subclasses of that module.
help() # This gives you some general examples to get you start with, as well as a link to the documentations.
```

As well as a tip that could potentially save you some times: STM32's pins' functions are predefined. Each pin has a fixed set of "alternative functions". You can view it [here](https://github.com/micropython/micropython/blob/master/ports/stm32/boards/stm32f411_af.csv). They can not do things that they are not assigned to on a hardware level. So you might find hard to use PWM (which is not in the machine module for a good reason) on `PC13`. Because that pin does not have an alternate functionality callled "Timer" short "TIM", and hence can not generate a PWM signal.

**Reply by mogplus88** [reply]
@nspsck  thanks for the explanation. I haven't used STM micros much, most of my fiddling has been with Arduinos, i.e. Atmel chips, and C. Things are a bit different with stm32 and PYthon.
;-)

**Reply by davefes** [reply]
15 years of Atmel, then first go with a Blackpill and Micropython 3 years ago.  Diverted onto ESP32 and PICO but thanks to @nspsck  and others back on to the Blackpill.  If you do any timing, ADC or low-power work I think any effort getting to grips with the Blackpill would be worthwhile.

**Reply by mogplus88** [reply]
Hah! I can't remember how long I've been playing with these things, maybe five years, on and off. I only really have one project anyway, a radio control transmitter for model aircraft. But my secret peccadillo is that, shock horror, I like programming. I started life as a Cobol programmer on mainframes in 1976, and spent most of my life around IT. I've played with ESP32 too (weird chip), and PICO. Had much fun trying to wrap my brain around PIO programming, but I finally got it to do what I wanted it to do. Now I'm looking forward to doing the same thing with the black pill. Just for fun!
;-)

**Reply by davefes** [reply]
Was having grief trying to get into DFU mode again, found this:
https://www.stm32duino.com/viewtopic.php?t=1680
and got in first try.  Only once so far. 

PA10 to ground instead of connecting PA9 to PA10.

**Reply by mogplus88** [reply]
I haven't tried dfu mode again, now that I have the latest version of MP flashed to the chip. Had other grief though, using the user button. Sometimes it works, sometimes it doesn't. There seems to be no rhyme or reason to it. Unplugging and plugging in again seems to make no difference. It will just randomly start working again. I had to solder the pins onto the board yesterday so I could use it in my breadboard, so maybe I cooked something. Dunno. Anyway I've got another one arriving in a few days. I'll be a bit more careful doing the soldering on that one. Got an ESP32-S3 coming too. Just for fun.
;-)
